### ISSUE TITLE
[QUESTION] CP loss reduction behavior in loss_func of pretrain_gpt.py

### ISSUE THREAD

**Issue by junjzhang** [NVIDIA/Megatron-LM#1124]
Hi, I wonder why the loss reduction of CP and DP is handled differently in `loss_func`，where CP is reduced on `loss`，and dp is reduced on detached `loss`. In my view, the loss reduction behavior for cross entropy is identical between CP and DP. Also, why should the output loss be multiplied by context_parallel_size?

```python
def loss_func(loss_mask: torch.Tensor, output_tensor: torch.Tensor):
    """Loss function.

    Args:
        loss_mask (torch.Tensor): Used to mask out some portions of the loss
        output_tensor (torch.Tensor): The tensor with the losses

    Returns:
        the loss scalar for this micro-batch
        the number of non-padded tokens in this microbatch
        a dict containing reporting metrics on the loss and number of tokens across
            the data parallel ranks
    """
    args = get_args()

    losses = output_tensor.float()
    loss_mask = loss_mask.view(-1).float()
    total_tokens = loss_mask.sum()
    loss = torch.cat([torch.sum(losses.view(-1) * loss_mask).view(1), total_tokens.view(1)])

    if args.context_parallel_size > 1:
        torch.distributed.all_reduce(loss, group=mpu.get_context_parallel_group())

    # Check individual rank losses are not NaN prior to DP all-reduce.
    if args.check_for_nan_in_loss_and_grad:
        global_rank = torch.distributed.get_rank()
        assert not loss[0].isnan(), (
            f'Rank {global_rank}: found NaN in local forward loss calculation. '
            f'Device: {torch.cuda.current_device()}, node: {os.uname()[1]}'
        )

    # Reduce loss for logging.
    reporting_loss = loss.clone().detach()
    torch.distributed.all_reduce(reporting_loss, group=mpu.get_data_parallel_group())

    local_num_tokens = loss[1].clone().detach().to(torch.int)
    return (
        loss[0] * args.context_parallel_size,
        local_num_tokens,
        {'lm loss': (reporting_loss[0], reporting_loss[1])},
    )
```

**Comment by heavyrain-lzy** [comment]
I also have this question. The previous issue https://github.com/NVIDIA/Megatron-LM/issues/673 explains the grads of `dp1cp2` and `dp2cp1` should be the same with the same global batch size. But I think the grads of `dp1cp2` and `dp1cp1` should also be the same with the same global batch size, but that's not the case.

**Comment by junjzhang** [comment]
> I also have this question. The previous issue [#673](https://github.com/NVIDIA/Megatron-LM/issues/673) explains the grads of `dp1cp2` and `dp2cp1` should be the same with the same global batch size. But I think the grads of `dp1cp2` and `dp1cp1` should also be the same with the same global batch size, but that's not the case.

Thanks for replying! I've checked that issue and that solved my problems. Also, megatron just [fixed a bug about grad scaling of CP in DDP](https://github.com/NVIDIA/Megatron-LM/commit/3bdcbbbe5d2a455a75e28969be7250cd4bd27bae), I think you should cherry-pick this one to your `FlagScale` repo.

**Comment by junjzhang** [comment]
#673

**Comment by junjzhang** [comment]
Same question as #673

**Comment by heavyrain-lzy** [comment]
> > I also have this question. The previous issue [#673](https://github.com/NVIDIA/Megatron-LM/issues/673) explains the grads of `dp1cp2` and `dp2cp1` should be the same with the same global batch size. But I think the grads of `dp1cp2` and `dp1cp1` should also be the same with the same global batch size, but that's not the case.
> 
> Thanks for replying! I've checked that issue and that solved my problems. Also, megatron just [fixed a bug about grad scaling of CP in DDP](https://github.com/NVIDIA/Megatron-LM/commit/3bdcbbbe5d2a455a75e28969be7250cd4bd27bae), I think you should cherry-pick this one to your `FlagScale` repo.

Maybe you should consider the `dp1cp1` and `dp1cp2` cases where the `num_microbachs` is the same.

**Comment by junjzhang** [comment]
> > > I also have this question. The previous issue [#673](https://github.com/NVIDIA/Megatron-LM/issues/673) explains the grads of `dp1cp2` and `dp2cp1` should be the same with the same global batch size. But I think the grads of `dp1cp2` and `dp1cp1` should also be the same with the same global batch size, but that's not the case.
> > 
> > 
> > Thanks for replying! I've checked that issue and that solved my problems. Also, megatron just [fixed a bug about grad scaling of CP in DDP](https://github.com/NVIDIA/Megatron-LM/commit/3bdcbbbe5d2a455a75e28969be7250cd4bd27bae), I think you should cherry-pick this one to your `FlagScale` repo.
> 
> Maybe you should consider the `dp1cp1` and `dp1cp2` cases where the `num_microbachs` is the same.

Yes, I've got the point that CP does not slice batch and the averaging is done by dividing total num tokens. I mean that Megatron just fixes the bug I mentioned above which indicates the grad should be divided by DP x CP size, and I noticed that [your repo also has this bug](https://github.com/FlagOpen/FlagScale/blob/237377e944e3935c868940b84c18b56c81fa10dd/megatron/megatron/core/distributed/distributed_data_parallel.py#L121). Maybe you should cherry pick this bugfix or sync with Megatron.
